kernel/src/snapshot.rs

```diff
@@ -67,6 +76,98 @@ impl Snapshot {
         Self::try_new_from_log_segment(table_root, log_segment, engine)
     }
 
+    /// Create a new [`Snapshot`] instance from an existing [`Snapshot`]. This is useful when you
+    /// already have a [`Snapshot`] lying around and want to do the minimal work to 'update' the
+    /// snapshot to a later version.
+    ///
+    /// # Parameters
+    ///
+    /// - `existing_snapshot`: reference to an existing [`Snapshot`]
+    /// - `engine`: Implementation of [`Engine`] apis.
+    /// - `version`: target version of the [`Snapshot`]. None will create a snapshot at the latest
+    ///   version of the table.
+    pub fn try_new_from(
+        existing_snapshot: Arc<Snapshot>,
+        engine: &dyn Engine,
+        version: impl Into<Option<Version>>,
+    ) -> DeltaResult<Arc<Self>> {
+        // simple heuristic for now:
+        // 1. if the new version == existing version, just return the existing snapshot
+        // 2. if the new version < existing version, just return an entirely new snapshot
+        // 3. list from (existing snapshot version + 1) onward
+        // 4a. if new checkpoint is found: just create a new snapshot from that checkpoint (and
+        // commits after it)
+        // 4b. if no new checkpoint is found: do lightweight P+M replay on the latest commits
+        let old_version = existing_snapshot.version();
+        let new_version = version.into();
+        if let Some(new_version) = new_version {
+            if new_version == old_version {
+                // Re-requesting the same version
+                return Ok(existing_snapshot.clone());
+            }
+            if new_version < old_version {
+                warn!(
+                    "[Snapshot::try_new_from] new version: {new_version} is older than existing version: {}",
+                    old_version
+                );
+                // Hint is too new, just create a new snapshot the normal way
+                return Snapshot::try_new(
+                    existing_snapshot.table_root().clone(),
+                    engine,
+                    Some(new_version),
+                )
+                .map(Arc::new);
+            }
+        }
+
+        // create a log segment just from existing_snapshot.version -> new_version
+        let log_root = existing_snapshot.log_segment.log_root.clone();
+        let fs_client = engine.get_file_system_client();
+        let new_log_segment = LogSegment::for_versions(
+            fs_client.as_ref(),
+            log_root,
+            existing_snapshot.version() + 1,
+            new_version,
+        );
+
+        match new_log_segment {
+            Err(Error::EmptyLogSegment) => {
```

Comment:
I don't _love_ errors as control flow. could `for_versions` return `DeltaResult<Option<LogSegment>>` and return `Ok(None)` if there is no log segment for the requested versions?

Reply:
removed this - doing something differently now